src/v/archival/async_data_uploader.cc

```diff
@@ -0,0 +1,348 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "archival/async_data_uploader.h"
+
+#include "archival/logger.h"
+#include "archival/types.h"
+#include "base/vlog.h"
+#include "model/record.h"
+#include "model/timeout_clock.h"
+#include "resource_mgmt/io_priority.h"
+#include "ssx/future-util.h"
+#include "storage/disk_log_impl.h"
+#include "storage/offset_to_filepos.h"
+#include "storage/record_batch_utils.h"
+#include "utils/retry_chain_node.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/shared_ptr.hh>
+
+#include <exception>
+#include <stdexcept>
+#include <variant>
+
+namespace archival {
+
+class reader_ds : public ss::data_source_impl {
+    /// Consumer that reads data batches to the read buffer
+    /// of the data source.
+    class consumer {
+    public:
+        explicit consumer(reader_ds* ds, inclusive_offset_range range)
+          : _parent(ds)
+          , _range(range) {}
+
+        /// This consumer accepts all batches if there is enough space in the
+        /// read buffer.
+        ss::future<ss::stop_iteration> operator()(model::record_batch batch) {
+            vlog(
+              _parent->_ctxlog.debug,
+              "consume started, buffered {} bytes",
+              _parent->_buffer.size_bytes());
+            if (!_range.contains(batch.header())) {
+                vlog(
+                  _parent->_ctxlog.debug,
+                  "skip batch {}-{}",
+                  batch.base_offset(),
+                  batch.last_offset());
+
+                co_return ss::stop_iteration::no;
+            }
+            vlog(
+              _parent->_ctxlog.debug,
+              "consuming batch {}-{}",
+              batch.base_offset(),
+              batch.last_offset());
+            auto hdr_iobuf = storage::batch_header_to_disk_iobuf(
+              batch.header());
+            auto rec_iobuf = std::move(batch).release_data();
+            _parent->_buffer.append(std::move(hdr_iobuf));
+            _parent->_buffer.append(std::move(rec_iobuf));
+            _parent->_num_batches++;
+            bool max_bytes_reached = _parent->_buffer.size_bytes()
+                                     > _parent->_max_bytes;
+            vlog(
+              _parent->_ctxlog.debug,
+              "max_bytes_reached: {}",
+              max_bytes_reached);
+            co_return max_bytes_reached ? ss::stop_iteration::yes
+                                        : ss::stop_iteration::no;
+        }
+
+        bool end_of_stream() const { return false; }
+
+    private:
+        reader_ds* _parent;
```

Comment:
nit: would prefer to see `reader_ds* _parent` as `reader_ds& _parent` since lifetimes are tied and `_parent` is non-nullable in this context.

Reply:
will fix in a followup